src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -74,11 +76,13 @@
 public abstract class SamReaderFactory {
 
     private static ValidationStringency defaultValidationStringency = ValidationStringency.DEFAULT_STRINGENCY;
-    
+
+    protected Function<SeekableByteChannel, SeekableByteChannel> pathWrapper;
+
     abstract public SamReader open(final File file);
 
     public SamReader open(final Path path) {
-        final SamInputResource r = SamInputResource.of(path);
+        final SamInputResource r = SamInputResource.of(path, getPathWrapper());
         final Path indexMaybe = SamFiles.findIndex(path);
         if (indexMaybe != null) r.index(indexMaybe);
```

Comment:
I just noticed, this isn't going to wrap the index.  That seems like a problem.  

We might actually want the ability to treat the index and the main path separately?  Do you think that's necessary?  I.e. download the index in it's entirety and store it in on disk vs cache chunks bam in memory as we stream over them.